### PR TITLE
Convert AWS API calls to use paginators

### PR DESCRIPTION
**Pull request by ungureanuvladvictor**
Some of the AWS API calls that we do are not doing correct pagination. This means that we return partial output to the calling functions. Those ones are (from the ec2 internal package):
- `describeVpcs`
- `describeSubnets`
- `describeSecurityGroups`

The ones that implement manually pagination are:
- `describeNetworkInterfaces`
- `GetInstanceTypes`

Overall it might not be important to implement pagination for all EC2 API calls but is good to standardize on it. Happy to revisit this if the reviewers think this is complicating things too much.

Signed-off-by: Vlad Ungureanu <vladu@palantir.com>

### CODE REVIEW

**Comment by ungureanuvladvictor** [comment]
test-me-please

File: pkg/policy/groups/aws/aws.go [hunk 1]
```diff
@@ -73,21 +81,21 @@ func getInstancesIpsFromFilter(ctx context.Context, filter *api.AWSGroup) ([]net
...
-	ec2Client := ec2.NewFromConfig(cfg)
-	result, err := ec2Client.DescribeInstances(ctx, input)
-	if err != nil {
-		return []net.IP{}, fmt.Errorf("Cannot retrieve aws information: %s", err)
+
+	paginator := ec2.NewDescribeInstancesPaginator(ec2Client, input)
+	for paginator.HasMorePages() {
+		output, err := paginator.NextPage(ctx)
+		if err != nil {
+			return []net.IP{}, fmt.Errorf("Cannot retrieve aws information: %s", err)
```
Comment (vadorovsky) [inline pkg/policy/groups/aws/aws.go]:
```suggestion
			return []net.IP{}, fmt.Errorf("Cannot retrieve aws information: %w", err)
```

**Review by vadorovsky — CHANGES_REQUESTED** [review]
One nit, the rest LGTM

**Comment by ungureanuvladvictor** [comment]
test-me-please

File: pkg/policy/groups/aws/aws.go (hunk 1 above)
Comment (ungureanuvladvictor) [reply pkg/policy/groups/aws/aws.go]: Updated!

File: pkg/aws/ec2/ec2.go [hunk 2]
```diff
@@ -118,48 +118,32 @@ func deriveStatus(err error) string {
...
-		interfacesFilters []ec2_types.Filter
-		nextToken         string
-	)
-
-	for {
-		c.limiter.Limit(ctx, "DescribeNetworkInterfaces")
-		input := &ec2.DescribeNetworkInterfacesInput{}
-		if nextToken != "" {
-			input.NextToken = &nextToken
+	result := []ec2_types.NetworkInterface{}
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: small nit, the slice doesn't need to be initialized, can be `var result []ec2_types.NetworkInterface`

File: pkg/aws/ec2/ec2.go [hunk 3]
```diff
@@ -257,20 +241,19 @@ func (c *Client) GetInstances(ctx context.Context, vpcs ipamTypes.VirtualNetwork
...
-	var vpcs []ec2_types.Vpc
-
-	c.limiter.Limit(ctx, "DescribeVpcs")
-
-	sinceStart := spanstat.Start()
-	output, err := c.ec2Client.DescribeVpcs(ctx, &ec2.DescribeVpcsInput{})
-	c.metricsAPI.ObserveAPICall("DescribeVpcs", deriveStatus(err), sinceStart.Seconds())
-	if err != nil {
-		return nil, err
+	result := []ec2_types.Vpc{}
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: same here for the initialization of the slice.

File: pkg/aws/ec2/ec2.go [hunk 4]
```diff
@@ -257,20 +241,19 @@ func (c *Client) GetInstances(ctx context.Context, vpcs ipamTypes.VirtualNetwork
...
+	paginator := ec2.NewDescribeVpcsPaginator(c.ec2Client, &ec2.DescribeVpcsInput{})
+	for paginator.HasMorePages() {
+		c.limiter.Limit(ctx, "DescribeVpcs")
+		sinceStart := spanstat.Start()
+		output, err := paginator.NextPage(ctx)
+		c.metricsAPI.ObserveAPICall("DescribeVpcs", deriveStatus(err), sinceStart.Seconds())
+		if err != nil {
+			return result, err
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: `return nil, err`

File: pkg/aws/ec2/ec2.go [hunk 5]
```diff
@@ -297,20 +280,24 @@ func (c *Client) GetVpcs(ctx context.Context) (ipamTypes.VirtualNetworkMap, erro
 
 // describeSubnets lists all subnets
 func (c *Client) describeSubnets(ctx context.Context) ([]ec2_types.Subnet, error) {
-	c.limiter.Limit(ctx, "DescribeSubnets")
-
-	sinceStart := spanstat.Start()
+	result := []ec2_types.Subnet{}
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: same here for the initialization of the slice.

File: pkg/aws/ec2/ec2.go [hunk 6]
```diff
@@ -494,15 +481,19 @@ func createAWSTagSlice(tags map[string]string) []ec2_types.Tag {
 }
 
 func (c *Client) describeSecurityGroups(ctx context.Context) ([]ec2_types.SecurityGroup, error) {
-	c.limiter.Limit(ctx, "DescribeSecurityGroups")
-	sinceStart := spanstat.Start()
-	output, err := c.ec2Client.DescribeSecurityGroups(ctx, &ec2.DescribeSecurityGroupsInput{})
-	c.metricsAPI.ObserveAPICall("DescribeSecurityGroups", deriveStatus(err), sinceStart.Seconds())
-	if err != nil {
-		return []ec2_types.SecurityGroup{}, err
+	result := []ec2_types.SecurityGroup{}
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: same here for the initialization of the slice.

File: pkg/aws/ec2/ec2.go [hunk 7]
```diff
@@ -494,15 +481,19 @@ func createAWSTagSlice(tags map[string]string) []ec2_types.Tag {
...
+	paginator := ec2.NewDescribeSecurityGroupsPaginator(c.ec2Client, &ec2.DescribeSecurityGroupsInput{})
+	for paginator.HasMorePages() {
+		c.limiter.Limit(ctx, "DescribeSecurityGroups")
+		sinceStart := spanstat.Start()
+		output, err := paginator.NextPage(ctx)
+		c.metricsAPI.ObserveAPICall("DescribeSecurityGroups", deriveStatus(err), sinceStart.Seconds())
+		if err != nil {
+			return result, err
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: `return nil, err`

File: pkg/aws/ec2/ec2.go [hunk 8]
```diff
@@ -536,28 +527,17 @@ func (c *Client) GetSecurityGroups(ctx context.Context) (types.SecurityGroupMap,
...
+	result := []ec2_types.InstanceTypeInfo{}
+	paginator := ec2.NewDescribeInstanceTypesPaginator(c.ec2Client, &ec2.DescribeInstanceTypesInput{})
+	for paginator.HasMorePages() {
+		c.limiter.Limit(ctx, "DescribeInstanceTypes")
+		sinceStart := spanstat.Start()
+		output, err := paginator.NextPage(ctx)
+		c.metricsAPI.ObserveAPICall("DescribeInstanceTypes", deriveStatus(err), sinceStart.Seconds())
 		if err != nil {
-			return instanceTypeInfos, err
+			return result, err
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: `return nil, err`

File: pkg/aws/ec2/ec2.go [hunk 9]
```diff
@@ -536,28 +527,17 @@ func (c *Client) GetSecurityGroups(ctx context.Context) (types.SecurityGroupMap,
...
-	}
-
-	instanceTypeInfos = append(instanceTypeInfos, output.InstanceTypes...)
-
-	for output.NextToken != nil {
-		describeInstanceTypes := &ec2.DescribeInstanceTypesInput{
-			NextToken: output.NextToken,
-		}
-		output, err = c.ec2Client.DescribeInstanceTypes(ctx, describeInstanceTypes)
+	result := []ec2_types.InstanceTypeInfo{}
```
Comment (aanm) [inline pkg/aws/ec2/ec2.go]: same here for the initialization of the slice.

File: pkg/policy/groups/aws/aws.go [hunk 10]
```diff
@@ -51,7 +51,15 @@ func GetIPsFromGroup(ctx context.Context, group *api.ToGroups) ([]net.IP, error)
 // getInstancesFromFilter returns the instances IPs in aws EC2 filter by the
 // given filter
 func getInstancesIpsFromFilter(ctx context.Context, filter *api.AWSGroup) ([]net.IP, error) {
+	result := []ec2_types.Reservation{}
```
Comment (aanm) [inline pkg/policy/groups/aws/aws.go]: same here for the initialization of the slice.

File: pkg/policy/groups/aws/aws.go [hunk 11]
```diff
@@ -51,7 +51,15 @@ func GetIPsFromGroup(ctx context.Context, group *api.ToGroups) ([]net.IP, error)
...
 	input := &ec2.DescribeInstancesInput{}
+
+	cfg, err := cilium_ec2.NewConfig(ctx)
+	if err != nil {
+		return []net.IP{}, err
```
Comment (aanm) [inline pkg/policy/groups/aws/aws.go]: `return nil, err`

File: pkg/policy/groups/aws/aws.go [hunk 12]
```diff
@@ -73,21 +81,21 @@ func getInstancesIpsFromFilter(ctx context.Context, filter *api.AWSGroup) ([]net
...
-	ec2Client := ec2.NewFromConfig(cfg)
-	result, err := ec2Client.DescribeInstances(ctx, input)
-	if err != nil {
-		return []net.IP{}, fmt.Errorf("Cannot retrieve aws information: %s", err)
+
+	paginator := ec2.NewDescribeInstancesPaginator(ec2Client, input)
+	for paginator.HasMorePages() {
+		output, err := paginator.NextPage(ctx)
+		if err != nil {
+			return []net.IP{}, fmt.Errorf("Cannot retrieve aws information: %w", err)
```
Comment (aanm) [inline pkg/policy/groups/aws/aws.go]: `return nil, fmt.Error...`

**Review by aanm — CHANGES_REQUESTED** [review]
A couple of small nits.

**Comment by ungureanuvladvictor** [comment]
test-me-please

File: pkg/aws/ec2/ec2.go (hunk 2 above)
Comment (ungureanuvladvictor) [reply pkg/aws/ec2/ec2.go]: Updated

File: pkg/aws/ec2/ec2.go (hunk 3 above)
Comment (ungureanuvladvictor) [reply pkg/aws/ec2/ec2.go]: Updated

File: pkg/aws/ec2/ec2.go (hunk 4 above)
Comment (ungureanuvladvictor) [reply pkg/aws/ec2/ec2.go]: Updated

File: pkg/aws/ec2/ec2.go (hunk 5 above)
Comment (ungureanuvladvictor) [reply pkg/aws/ec2/ec2.go]: Updated

File: pkg/aws/ec2/ec2.go (hunk 6 above)
Comment (ungureanuvladvictor) [reply pkg/aws/ec2/ec2.go]: Updated

File: pkg/aws/ec2/ec2.go (hunk 8 above)
Comment (ungureanuvladvictor) [reply pkg/aws/ec2/ec2.go]: Updated

File: pkg/aws/ec2/ec2.go (hunk 9 above)
Comment (ungureanuvladvictor) [reply pkg/aws/ec2/ec2.go]: Updated

File: pkg/policy/groups/aws/aws.go (hunk 10 above)
Comment (ungureanuvladvictor) [reply pkg/policy/groups/aws/aws.go]: Updated

File: pkg/policy/groups/aws/aws.go (hunk 12 above)
Comment (ungureanuvladvictor) [reply pkg/policy/groups/aws/aws.go]: Updated

File: pkg/policy/groups/aws/aws.go (hunk 11 above)
Comment (ungureanuvladvictor) [reply pkg/policy/groups/aws/aws.go]: Updated

**Review by aanm — APPROVED** [review]

**Comment by ungureanuvladvictor** [comment]
retest-net-next

**Review by christarazi — APPROVED** [review]
LGTM. Can you put the context in the PR into the commit itself? It would be helpful to have the "why" in the commit msg

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Pull request merged**